File: Leetcode/06_Stack & Queue/033_0155_min-stack.py
class MinStack:
    
    def __init__(self):
        self.items = []
        self.minstack = []
        

    def push(self, val: int):
        self.items.append(val)
        if len(self.minstack) == 0 or val <= self.minstack[-1]:
            self.minstack.append(val)
        

    def pop(self):
        if len(self.items) == 0:
            return
        top = self.items[-1]
        if len(self.minstack) > 0 and top == self.minstack[-1]:
            self.minstack.pop()
        self.items.pop()
        

    def top(self):
        return self.items[-1]
        

    def getMin(self):
        if len(self.minstack) == 0:
            return
        return self.minstack[-1]
        


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: Leetcode/06_Stack & Queue/033_0155_min-stack.go
// https://leetcode.cn/problems/min-stack/description/
// m

package main

type MinStack struct {
	items    []int
	minStack []int
}

func Constructor() MinStack {
	return MinStack{}
}

func (s *MinStack) Push(val int) {
	s.items = append(s.items, val)
	if len(s.minStack) == 0 || val <= s.minStack[len(s.minStack)-1] {
		s.minStack = append(s.minStack, val)
	}
}

func (s *MinStack) Pop() {
	if len(s.items) == 0 {
		return
	}
	top := s.items[len(s.items)-1]
	s.items = s.items[:len(s.items)-1]
	if top == s.minStack[len(s.minStack)-1] {
		s.minStack = s.minStack[:len(s.minStack)-1]
	}
}

func (s *MinStack) Top() int {
	return s.items[len(s.items)-1]
}

func (s *MinStack) GetMin() int {
	return s.minStack[len(s.minStack)-1]
}
